Source/DafnyLanguageServer/Workspace/CompilationManager.cs

```diff
@@ -0,0 +1,346 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using VC;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace;
+
+/// <summary>
+/// Manages the compilation of a single document version.
+/// Agnostic to document updates. Does not handle migration of old document state.
+/// </summary>
+public class CompilationManager {
+
+  private readonly ILogger logger;
+  private readonly ITextDocumentLoader documentLoader;
+  private readonly ICompilationStatusNotificationPublisher statusPublisher;
+  private readonly INotificationPublisher notificationPublisher;
+  private readonly IProgramVerifier verifier;
+
+  private DafnyOptions Options => DafnyOptions.O;
+  private VerifierOptions VerifierOptions { get; }
+
+  public DocumentTextBuffer TextBuffer { get; }
+  private readonly Range? lastChange;
+  private readonly IServiceProvider services;
+  private readonly ImmutableList<Position> migratedLastTouchedVerifiables;
+  private bool shouldVerify;
+  private readonly VerificationTree? migratedVerificationTree;
+
+  private readonly IScheduler verificationUpdateScheduler = new EventLoopScheduler();
+  private readonly CancellationTokenSource cancellationSource;
+  private readonly ReplaySubject<DafnyDocument> documentUpdates = new();
+  public IObservable<DafnyDocument> DocumentUpdates => documentUpdates;
+
+  public Task<DafnyDocument> ResolvedDocument { get; }
+  public Task<DafnyDocument> TranslatedDocument { get; }
+
+  public CompilationManager(IServiceProvider services,
+    VerifierOptions verifierOptions,
+    DocumentTextBuffer textBuffer,
+    Range? lastChange,
+    bool shouldVerify,
+    ImmutableList<Position> migratedLastTouchedVerifiables,
+    VerificationTree? migratedVerificationTree) {
+    VerifierOptions = verifierOptions;
+    logger = services.GetRequiredService<ILogger<CompilationManager>>();
+    documentLoader = services.GetRequiredService<ITextDocumentLoader>();
+    notificationPublisher = services.GetRequiredService<INotificationPublisher>();
+    verifier = services.GetRequiredService<IProgramVerifier>();
+    statusPublisher = services.GetRequiredService<ICompilationStatusNotificationPublisher>();
+
+    TextBuffer = textBuffer;
+    this.services = services;
+    this.migratedLastTouchedVerifiables = migratedLastTouchedVerifiables;
+    this.shouldVerify = shouldVerify;
+    this.migratedVerificationTree = migratedVerificationTree;
+    this.lastChange = lastChange;
+    cancellationSource = new();
+
+    ResolvedDocument = ResolveAsync();
+    TranslatedDocument = TranslateAsync();
+
+    if (shouldVerify) {
+      var _ = VerifyEverythingAsync();
+    } else {
+      MarkVerificationFinished();
+    }
+  }
+
+  public void Verify() {
+    if (shouldVerify) {
+      return;
+    }
+
+    shouldVerify = true;
```

Comment:
There can't be a race condition on this because this class is used from a single thread; correct?  It might be good to document it as not being thread-safe?